Add piv build dependencies

Joerger opened the pull request:
[piv-go](https://github.com/ericchiang/piv-go#installation) on Linux depends on [PCSC lite](https://pcsclite.apdu.fr/). This PR introduces this dependency and statically links it in releases.

Prerequisite for https://github.com/gravitational/teleport/pull/15874 and https://github.com/gravitational/teleport/pull/15335

Changes:
  - Add `LIBPCSCLITE_BUILD_TAG` and `LIBPCSCLITE_TEST_TAG`.
  - Add libpcsclite static linking, so that releases do not depend on the dynamic `libpcsclite.so` library.
    - This is only done for Centos7 (`build.assets/Dockerfile-centos7`) because as I understand from https://github.com/gravitational/teleport/pull/14062, we now build our releases with Centos7.
    - `libpcsclite-dev` remains in `build.assets/Dockerfile` for tests, but other pcsc dependencies added in https://github.com/gravitational/teleport/pull/16083/files have been removed in favor of the static version.
    - `LIBPCSCLITE=dynamic` can be used to link to the dynamic `libpcsclite.so` library (local builds)

e PR - https://github.com/gravitational/teleport.e/pull/526

Joerger committed:
Add piv build dependencies.

  - Add LIBPCSCLITE build tag.

  - Add libpcsclite static linking.

  - Refactor CGOFLAG in Makefile.

  - Update Centos7 Dockerfile and drone.

Co-authored-by: Jakub Nyckowski <jakub.nyckowski@goteleport.com>

zmb3 commented on build.assets/Dockerfile-centos7:
```diff
@@ -92,6 +92,32 @@ RUN mkdir -p /opt && cd /opt && \
+
+# Install libpcsclite.
+# Depends on autoconf.
+# We use outdated version 1.8.11 for licence compatibility.
```
```suggestion
# We use outdated version 1.8.11 for license compatibility.
```

tigrato reviewed (approved):
LGTM

Joerger committed:
Refactor RELEASE_MESSAGE for readability.

Now produces message like: "RELEASE_MESSAGE=Building with GOOS=linux GOARCH=amd64 REPRODUCIBLE= and with PIV support and without PAM support, FIPS support, BPF support, Windows RDP client, libfido2, Touch ID."

Joerger committed:
Add LIBPCSCLITE_VERSION docker arg.

Joerger committed:
Enable use of dynamic pcsc library with LIBPCSCLITE=dynamic.

Joerger committed:
Update e ref.

Joerger committed:
build PCSCLite from gravitational fork.

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies

jakule commented on Makefile:
```diff
@@ -35,63 +35,36 @@ TELEPORT_DEBUG ?= false
 
 # FIPS support must be requested at build time.
-FIPS_MESSAGE := "without FIPS support"
+FIPS_MESSAGE := without-FIPS-support
```
Why did you add `-` here?

Joerger replied:
This commit 7937f9824f55858cbd20c5bd4e2b0962e3c58e0c parses the messages with dashes to produce a more readable build message. This was in response to this comment - https://github.com/gravitational/teleport/pull/15335#discussion_r961241934

jakule commented on build.assets/Dockerfile-centos7:
```diff
@@ -92,6 +92,31 @@ RUN mkdir -p /opt && cd /opt && \
+
+# Install libpcsclite - compile with a newer GCC. The one installed by default is not able to compile it.
+ARG LIBPCSCLITE_VERSION
+RUN git clone --depth=1 https://github.com/gravitational/PCSC.git && \
```
Can you create a tag in our fork and pin the version here? Otherwise, we won't know which version was used for each build.

jakule commented on build.assets/Makefile:
```diff
@@ -36,6 +36,7 @@ NODE_VERSION ?= 16.13.2
 RUST_VERSION ?= 1.63.0 # don't bump this without checking GLIBC compatibility
 BORINGCRYPTO_RUNTIME=$(GOLANG_VERSION)b7
 LIBBPF_VERSION ?= 0.7.0-teleport
+LIBPCSCLITE_VERSION ?= 1.9.8
```
Can you rename the tag to `1.9.8-teleport` to indicate that we're not using the mainstream version? The same as we did with `libbpf` line above.

jakule commented on Makefile:
```diff
@@ -164,24 +136,44 @@ endif
 TOUCHID_TAG := touchid
 endif
 
+# Enable libpcsclite for testing?
```
Why does the "testing" build links dynamically and "non-testing" statically? If we want to avoid this dependency, we could just disable the tag from the build to not force everyone to install it.

Joerger replied:
The `LIBPCSCLITE_TEST_TAG` build tag is only used by `make test-go` and `make -C build.assets test` to dynamically link to `libpcsclite` for tests. `LIBPCSCLITE_TEST_TAG` does not introduce the dependency when building binaries. This is the same logic used for `libfido2` with the `LIBFIDO2_TEST_TAG` tag.

This isn't necessary but in my opinion it's a nice piece of helper logic to test `libpcsclite` when dynamic linking is possible and static linking is not set up, primarily for local developement. Let me know if I'm missing something though.

jakule commented on Makefile:
```diff
@@ -164,24 +136,44 @@ endif
+# Only build if LIBPCSCLITE=yes.
+# This is used for PIV functionality.
+LIBPCSCLITE_MESSAGE := without-PIV-support
+ifneq (, $(filter $(LIBPCSCLITE), yes static dynamic))
```
Why are you checking the message here? You could check the `LIBPCSCLITE_BUILD_TAG` here and check if it's set already.

Joerger committed:
link 1.9.9-teleport tag.

Joerger committed:
Merge branch 'joerger/add-piv-dependencies' of github.com:gravitational/teleport into joerger/add-piv-dependencies

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies

Joerger committed:
Add comments for testing tag, and dynamic vs static build tag.

Joerger committed:
Update Makefile

Co-authored-by: Jakub Nyckowski <jakub.nyckowski@goteleport.com>

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies

Joerger committed:
Update e ref.

Joerger committed:
Merge branch 'joerger/add-piv-dependencies' of github.com:gravitational/teleport into joerger/add-piv-dependencies

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies

Joerger committed:
Merge branch 'master' into joerger/add-piv-dependencies
